darshan-runtime/lib/darshan-pnetcdf-api.m4

```diff
@@ -0,0 +1,760 @@
+            }
+            DARSHAN_TIMER_INC_NO_OVERLAP(
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_READ_TIME],
+                tm1, tm2, rec_ref->last_read_end);',
+`$1',`put',`rec_ref->var_rec->counters[PNETCDF_VAR_BYTES_WRITTEN] += $3;
+            rec_ref->var_rec->counters[`PNETCDF_VAR_PUT_VAR'Upcase($2)] += 1;
+            if (rec_ref->last_io_type == DARSHAN_IO_READ)
+                rec_ref->var_rec->counters[PNETCDF_VAR_RW_SWITCHES] += 1;
+            rec_ref->last_io_type = DARSHAN_IO_WRITE;
+            DARSHAN_BUCKET_INC(
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_SIZE_WRITE_AGG_0_100]), $3);
+            common_access_vals[0] = $3;
+            cvc = darshan_track_common_val_counters(&rec_ref->access_root,
+                common_access_vals, PNETCDF_VAR_MAX_NDIMS+PNETCDF_VAR_MAX_NDIMS+1, &rec_ref->access_count);
+            if (cvc) DARSHAN_UPDATE_COMMON_VAL_COUNTERS(
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_ACCESS1_ACCESS]),
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_ACCESS1_COUNT]),
+                cvc->vals, cvc->nvals, cvc->freq, 0);
+            if (rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] == 0 ||
+             rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] > tm1)
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] = tm1;
+            rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_END_TIMESTAMP] = tm2;
+            if (rec_ref->var_rec->fcounters[PNETCDF_VAR_F_MAX_WRITE_TIME] < tm2 - tm1) {
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_MAX_WRITE_TIME] = tm2 - tm1;
+                rec_ref->var_rec->counters[PNETCDF_VAR_MAX_WRITE_TIME_SIZE] = $3;
+            }
+            DARSHAN_TIMER_INC_NO_OVERLAP(
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_TIME],
+                tm1, tm2, rec_ref->last_write_end);')')dnl
+dnl
+dnl
+define(`UPDATE_INDEPCOLL_RW_COUNTER',`ifelse(
+       `$1$2',`get',`rec_ref->var_rec->counters[PNETCDF_VAR_INDEP_READS] += 1',
+       `$1$2',`get_all',`rec_ref->var_rec->counters[PNETCDF_VAR_COLL_READS] += 1',
+       `$1$2',`put',`rec_ref->var_rec->counters[PNETCDF_VAR_INDEP_WRITES] += 1',
+       `$1$2',`put_all',`rec_ref->var_rec->counters[PNETCDF_VAR_COLL_WRITES] += 1')')dnl
+dnl
+
+dnl
+define(`APINAME',`ifelse(`$3',`',`ncmpi_$1_var$2$4',`ncmpi_$1_var$2$3$4')')dnl
+dnl
+
+/* ncmpi_get/put_var<kind>_<type>_<mode> API */
+dnl
+dnl GETPUT_API(get/put, `'/1/a/s/m, `'/itype, `'/_all)
+dnl
+define(`GETPUT_API',`
+DARSHAN_FORWARD_DECL(APINAME($1,$2,$3,$4), int, (int ncid, int varid, ArgKind($2) BufArgs($1,$3)));
+
+int DARSHAN_DECL(APINAME($1,$2,$3,$4))(int ncid, int varid, ArgKind($2)BufArgs($1,$3))
+{
+    int err, ret;
+    MPI_Offset $1_before, $1_after;
+    double tm1, tm2;
+
+    MAP_OR_FAIL(APINAME($1,$2,$3,$4));
+
+    err = ncmpi_inq_$1_size(ncid, &$1_before);
```

Comment:
Yes, ncmpi_inq_put_size and ncmpi_inq_get_size are low overhead.

The only metadata operation that can happen during the ncmpi_put/get_varxxx
calls is to update the time records, which is a 4 or 8 byte integer in the file header, if the variable has a time dimension.

Reply:
Good to know on the overhead front.

Reply:
I'm mostly just worried about a semantic mismatch with the PnetCDF module and HDF5 in terms of measuring the total size of a read/write access. In the HDF5 module, access size information only contains application dataset data, not any internal library metadata. If we can help it, I think it'd be nice to use a similar semantic here, even if it's only a small potential difference per operation. Any idea how involved it would be to define macros to extract the raw variable I/O buffer sizes from the given arguments in these functions?

I think knowing how much metadata is written is additionally useful, but we are already capturing this at the file level using these same calls.

Reply:
How does HDF5 provide the access size information? Or is this calculated by Darshan using filespace property? Is it the I/O amount after type conversion? for example, the app writes a buffer of double type in memory to a dataset of float type in file.

PnetCDF's inq_get/put_size APIs are designed to return the true I/O amount to the file system, including file metadata, with the type conversion case considered. One way to detect whether a write request updates the time record in the file header is to query the dimension size before and after. If the size increases, then metadata amount is 4 (for NC_64BIT_OFFSET format) or 8 (for NC_64BIT_DATA format).

Reply:
HDF5 access information is captured using the number of points selected in the file data space times the type size associated with the dataset. So, I guess that would be after the type conversion (but before operations like compression, etc.)?

There are a couple of different approaches you can take in what to count as an access size: what users see at the API level, and what the library passes through to the file system. I'm not necessarily taking the position one is right or wrong, just that I think it creates problems for us in conveying semantics of various counters, when they could vary from the PnetCDF module to the HDF5 module. Darshan analysis tools are unlikely to reliably account for that.

Reply:
The underlying metadata operations that might happen internal to HDF5/PnetCDF will be captured at the MPI-IO and POSIX layers by Darshan, so we aren't completely missing the I/O. If I as as user of the PnetCDF API ask it to write 8 bytes, and it writes 12 instead (4 bytes of metadata), I just don't think the PnetCDF module should count that as a 12-byte access. I think this is common practice for calculating attained bandwidth for high-level I/O libraries -- you calculate the total amount of data the application read/write divided by the time, you don't include any additional metadata written by the library itself.

Reply:
I am OK if you want to exclude any metadata.

I believe getting the total I/O amount for an I/O request has an advantage.
Often times, we are seeing and wondering why two requests with the same
amount yields different timings. If it turns out there are additional metadata
I/O activities happening underneath, then this can be a valuable information.

Reply:
@wkliao I was thinking more about how to calculate the total put/get size based on input parameters to I/O calls rather than using the pnetcdf routines discussed above, and wanted to see if you had feedback. The idea is that we have different ways of calculating how many "values" will be read/written depending on the API variant and then multiplying those number of values by the file external type size (which we already capture and store in PNETCDF_VAR module) to get a total I/O size for the operation. The number of values would be calculated like this for different API variants:

 - var1: 1 value
 - var: all values
 	- for variables with an unlimited dimension, this would need to be calculated; otherwise, the number of values is already captured by the PNETCDF_VAR module
 - vara: array of values
 	- multiply all input counts to get total
 - vars: subsampled array,
 	- multiply all input counts to get total
 - varm: mapped array
 	- multiply all input counts to get total
 - varn: list of subarrays
 	- multiply all input counts across all requests to get total
 - vard: MPI derived datatype
	 - use MPI_Type_size to get type size of input MPI datatype

I think the only snag is for the plain var API we would need some special handling to calculate the current size of a variable that has an unlimited dimension, since it's not a static value. It looks like I could use `ncmpi_inq_vardimid()` and `ncmpi_inq_dimlen()` functions to help with this though?

Does that make sense?

Reply:
That sounds right, if not using `ncmpi_inq_$1_size`.
As for var API, we can capture the variable size information at `ncmpi_def_var`,
but inquiring current unlimited dimension size is still needed.
https://github.com/darshan-hpc/darshan/blob/38ce1f401aa3cafb5ac90d31eb1181614ec0adcb/darshan-runtime/lib/darshan-pnetcdf-api.m4#L712

Reply:
I had a question after reading the following line in the PnetCDF docs related to `get/put_vara`:
```
count

    A vector of MPI_Offset integers specifying the edge lengths along each dimension of the block of data values to be written. To write a single value, for example, specify count as (1, 1, ... , 1). The length of count is the number of dimensions of the specified variable. The elements of count correspond to the variable’s dimensions. Hence, if the variable is a record variable, the first element of count corresponds to a count of the number of records to write.
```

As I understand, a "record variable" is just a variable that has an unlimited dimension. The last sentence above seems to imply that the total number of elements to write for record variables is given by the count value in only the first dimension (i.e., the unlimited dimension). But, I thought the total count would be calculated by multiplying the count from all dimensions, just as we would do for non-record variables. Am I misunderstanding the comment, or am I misunderstanding how the total number of records to update is counted for these interfaces?

Reply:
A `record` here is the subarray of a variable with the 1st dimension equal to 1.

Reply:
I see, I think I understand the terminology in the docs a little better with subtlety explained. To make sure I'm fully clear, it sounds like the thing to do is to always multiply count value from all dimensions, whether the variable is a record variable or not?

Reply:
Yes.

Reply:
@wkliao I added support for this in [2edfbc7](https://github.com/darshan-hpc/darshan/pull/456/commits/2edfbc71d7f52eff0021915be6d30e9a48ad7d39).

I basically just added a macro to calculate total access size in terms of input arguments to the various API variants and removed the calls to `get/put_size()`. I spot checked the Darshan log output for various test cases that touch different APIs (e.g., put_vara for `vara` interface, transpose for `vara` and `varm` interface, put_varn_float for `varn` interface, and vard_int for `vard` interface), and things mostly match. As you pointed out initially, in cases where there is a time variable, there are discrepancies since the PnetCDF APIs incur some additional metadata I/O:

original code:
```
PNETCDF_VAR     -1      4405928841334190961     PNETCDF_VAR_BYTES_READ  288     /home/shane/software/pnetcdf/pnetcdf-1.12.3/build/examples/C/time_var.nc:rec_var        /home   ext4
PNETCDF_VAR     -1      4405928841334190961     PNETCDF_VAR_BYTES_WRITTEN       296     /home/shane/software/pnetcdf/pnetcdf-1.12.3/build/examples/C/time_var.nc:rec_var        /home   ext4
```
my updates:
```
PNETCDF_VAR     -1      4405928841334190961     PNETCDF_VAR_BYTES_READ  288     /home/shane/software/pnetcdf/pnetcdf-1.12.3/build/examples/C/time_var.nc:rec_var        /home   ext4
PNETCDF_VAR     -1      4405928841334190961     PNETCDF_VAR_BYTES_WRITTEN       288     /home/shane/software/pnetcdf/pnetcdf-1.12.3/build/examples/C/time_var.nc:rec_var        /home   ext4
```

Feel free to provide any feedback on the changes. The `m4` code is admittedly hard to read, so might be easier to scan through the generated source to see what code was generated for each variant if you do want a closer look.

Reply:
I'm going to mark this as resolved. This new macro for calculating access size information is useful for reasons discussed earlier in this thread, but it's also the only method we have for instrumenting per-variable access size info for non-blocking APIs.